src/language-js/needs-parens.js

```diff
@@ -745,6 +755,14 @@ function needsParens(path, options) {
         parent.type !== "TypeCastExpression" &&
         parent.type !== "VariableDeclarator"
       );
+    case "TupleTypeAnnotation":
+    case "ObjectTypeAnnotation": {
+      const grandParent = path.getParentNode(1);
+      return (
+        grandParent.type === "ArrowFunctionExpression" &&
+        isIncludeFunctionTypeInObjectType(path)
+      );
+    }
```

Comment:
Instead of modifying all these `case`s, let's add `case "TypeAnnotation"`:
```js
    // workaround for https://github.com/facebook/flow/pull/8163
    case "TypeAnnotation":
      return (
        name === "returnType" &&
        parent.type === "ArrowFunctionExpression" &&
        includesFunctionTypeInObjectType(path)
      );
```

Reply:
One more snippet to test:
```
var example = (): (?{ p: string => string }) => (0: any);
```

Reply:
Please add this snippet to the test.